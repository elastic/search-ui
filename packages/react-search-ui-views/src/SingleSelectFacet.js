import PropTypes from "prop-types";
import React from "react";
import Select, { components } from "react-select";

import { FacetValue } from "./types";
import { getFilterValueDisplay } from "./view-helpers";
import { appendClassName } from "./view-helpers";

function Option(props) {
  return (
    <components.Option {...props}>
      <span className="sui-select__option-label">{props.data.label}</span>
      <span className="sui-select__option-count">
        {props.data.count.toLocaleString("en")}
      </span>
    </components.Option>
  );
}

Option.propTypes = {
  data: PropTypes.object.isRequired
};

function toSelectBoxOption(filterValue) {
  return {
    value: filterValue.value,
    label: getFilterValueDisplay(filterValue.value),
    count: filterValue.count
  };
}

const setDefaultStyle = {
  option: () => ({}),
  control: () => ({}),
  dropdownIndicator: () => ({}),
  indicatorSeparator: () => ({})
};

function SingleSelectFacet({ className, label, onChange, options }) {
  let selectedSelectBoxOption;
  let isSelectedSelectBoxOptionSet = false;

  const selectBoxOptions = options.map(option => {
    const selectBoxOption = toSelectBoxOption(option);
    // There should never be multiple filters set for this facet because it is single select,
    // but if there is, we use the first value.
    if (option.selected && !isSelectedSelectBoxOptionSet) {
      selectedSelectBoxOption = selectBoxOption;
      isSelectedSelectBoxOptionSet = true;
    }
    return selectBoxOption;
  });

  return (
    <div className={appendClassName("sui-facet", className)}>
      <div className="sui-facet__title">{label}</div>
      <Select
        className="sui-select"
        classNamePrefix="sui-select"
        components={{ Option }}
        value={selectedSelectBoxOption}
        onChange={o => onChange(o.value)}
        options={selectBoxOptions}
        isSearchable={false}
        styles={setDefaultStyle}
      />
    </div>
  );
}

SingleSelectFacet.propTypes = {
  label: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  options: PropTypes.arrayOf(FacetValue).isRequired,
  className: PropTypes.string
};

export default SingleSelectFacet;
