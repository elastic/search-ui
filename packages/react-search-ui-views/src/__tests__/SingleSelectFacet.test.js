import React from "react";
import SingleSelectFacet from "../SingleSelectFacet";
import { shallow, render } from "enzyme";

const params = {
  label: "A Facet",
  onChange: jest.fn(),
  options: [
    {
      count: 20,
      value: {
        from: 1,
        to: 10,
        name: "Range 1"
      },
      selected: true
    },
    {
      count: 10,
      value: {
        to: 20,
        from: 11,
        name: "Range 2"
      },
      selected: false
    }
  ]
};

it("renders", () => {
  const wrapper = shallow(<SingleSelectFacet {...params} />);
  expect(wrapper).toMatchSnapshot();
});

it("renders falsey values correctly", () => {
  const wrapper = shallow(
    <SingleSelectFacet
      {...params}
      options={[
        {
          value: 0,
          count: 10,
          selected: false
        },
        {
          value: false,
          count: 20,
          selected: false
        },
        {
          value: "",
          count: 30,
          selected: false
        }
      ]}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

describe("determining selected option", () => {
  it("will correctly determine which of the options is selected", () => {
    const wrapper = render(<SingleSelectFacet {...params} />);
    expect(wrapper.find(".sui-select__single-value").text()).toEqual("Range 1");
  });

  // This shouldn't ever happen, but if it does, it should use the first selected value
  it("will used the first selected option when multiple options are selected", () => {
    const wrapper = render(
      <SingleSelectFacet
        {...params}
        options={params.options.map(o => ({ ...o, selected: true }))}
      />
    );
    expect(wrapper.find(".sui-select__single-value").text()).toEqual("Range 1");
  });

  it("will correctly determine when no value is selected", () => {
    const wrapper = render(
      <SingleSelectFacet
        {...params}
        options={params.options.map(o => ({ ...o, selected: false }))}
      />
    );
    expect(wrapper.find(".sui-select__single-value").text()).toEqual("");
  });
});

it("renders with className prop applied", () => {
  const customClassName = "test-class";
  const wrapper = shallow(
    <SingleSelectFacet {...params} className={customClassName} />
  );
  const { className } = wrapper.props();
  expect(className).toEqual("sui-facet test-class");
});
