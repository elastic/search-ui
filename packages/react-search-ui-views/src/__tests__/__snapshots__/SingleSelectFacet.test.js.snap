// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="sui-facet"
>
  <div
    className="sui-facet__title"
  >
    A Facet
  </div>
  <StateManager
    className="sui-select"
    classNamePrefix="sui-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "count": 20,
          "label": "Range 1",
          "value": Object {
            "from": 1,
            "name": "Range 1",
            "to": 10,
          },
        },
        Object {
          "count": 10,
          "label": "Range 2",
          "value": Object {
            "from": 11,
            "name": "Range 2",
            "to": 20,
          },
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
      }
    }
    value={
      Object {
        "count": 20,
        "label": "Range 1",
        "value": Object {
          "from": 1,
          "name": "Range 1",
          "to": 10,
        },
      }
    }
  />
</div>
`;

exports[`renders falsey values correctly 1`] = `
<div
  className="sui-facet"
>
  <div
    className="sui-facet__title"
  >
    A Facet
  </div>
  <StateManager
    className="sui-select"
    classNamePrefix="sui-select"
    components={
      Object {
        "Option": [Function],
      }
    }
    defaultInputValue=""
    defaultMenuIsOpen={false}
    defaultValue={null}
    isSearchable={false}
    onChange={[Function]}
    options={
      Array [
        Object {
          "count": 10,
          "label": "0",
          "value": 0,
        },
        Object {
          "count": 20,
          "label": "false",
          "value": false,
        },
        Object {
          "count": 30,
          "label": "",
          "value": "",
        },
      ]
    }
    styles={
      Object {
        "control": [Function],
        "dropdownIndicator": [Function],
        "indicatorSeparator": [Function],
        "option": [Function],
      }
    }
  />
</div>
`;
